src/Chapter3.hs

```diff
@@ -480,6 +552,80 @@ After defining the city, implement the following functions:
    and at least 10 living __people__ inside in all houses of the city totally.
 -}
 
+data Castle = Castle
+    { name :: String
+    } deriving (Show)
+
+data Wall = Wall deriving (Show)
+data Church = Church deriving (Show)
+data Library = Library deriving (Show)
+data People
+    = OnePerson
+    | TwoPersons
+    | ThreePersons
+    | FourPersons
+    deriving (Show)
+
+data House = House
+    { people :: People
+    } deriving (Show)
+
+data City
+    = CityWithCastleWallChurch
+        { castle  :: Castle
+        , walls    :: [Wall]
+        , church  :: Church
+        , houses  :: [House]
+        }
+    | CityWithCastleWallLibrary
+        { castle  :: Castle
+        , wall    :: [Wall]
+        , library :: Library
+        , houses  :: [House]
+        }
+    | CityWithChurch
+        { church  :: Church
+        , houses  :: [House]
+        }
+    | CityWithLibrary
+        { library :: Library
+        , houses  :: [House]
+        }
+    deriving (Show)
+
+buildCastle :: City -> String -> City
+buildCastle city castleName = city { castle = Castle castleName}
+
+countPeople :: House -> Int
+countPeople house = case people house of
+    OnePerson -> 1
+    TwoPersons -> 2
+    ThreePersons -> 3
+    FourPersons -> 4
+
+buildHouse  :: City -> Int -> City
+buildHouse city numPersons = city { houses = (new_house:houses city)}
+    where
+        new_house
+            | numPersons == 1 = House OnePerson
+            | numPersons == 2 = House TwoPersons
+            | numPersons == 3 = House ThreePersons
+            | numPersons == 4 = House FourPersons
+            | otherwise = error "Invalid number of persons"
```

Comment:
Nice smart constructor in here! 👍🏼 

Note that you could use the `case-of` construction in here:

```haskell
        newHouse = case numPersons of
            1 -> House OnePerson
...
```

It is usually not that good to have `error` in code. It is better to use more typesafe interfaces. 
For example in here you can create the function like this:

```haskell
makeHouse :: Int -> Maybe House
makeHouse numPersons = case numPersons of
    1 -> Just (House OnePerson)
    2 -> ...
    3 -> ...
    4 -> ...
    _ -> Nothing
```

And in `buildHouse` you can pattern match on the result of `makeHouse` and add the house only in case of `Just`. In case of `Nothing you could return unchanged city 🙂